### ISSUE TITLE
Broken QuarkusTestExtension

### ISSUE THREAD

**Issue by alesj** [quarkusio/quarkus#8446]
When using Quarkus 1.3.1.Final or current master against our Registry, I get this:
* https://gist.githubusercontent.com/alesj/86aae892c3aa35113ce358737700b4b7/raw/8a3e105e971916bd2f997692d145043d5810ee2a/gistfile1.txt

Zulip thread:
* https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/Broken.20QuarkusTestExtension.20.2E.2E.2E

As mentioned / described, we use custom JUnit5 extension
* https://github.com/Apicurio/apicurio-registry/blob/master/utils/tests/src/main/java/io/apicurio/registry/utils/tests/RegistryServiceExtension.java
(it's similar to existing JUnit5 Parameter extension, just custom for RegistryService)

To reproduce, simply change Quarkus version here:
* https://github.com/Apicurio/apicurio-registry
and run "mvn clean install"

**Comment by alesj** [comment]
newMethod = {java.lang.reflect.Method@8575} "void io.apicurio.registry.ArtifactStateTest.testEnableDisableArtifact(io.apicurio.registry.client.RegistryService)"
 clazz = {java.lang.Class@4485} "class io.apicurio.registry.ArtifactStateTest"
 slot = 5
 name = "testEnableDisableArtifact"
 returnType = {java.lang.Class@8581} "void"
 parameterTypes = {java.lang.Class[1]@8582} 
  0 = {java.lang.Class@8475} "interface io.apicurio.registry.client.RegistryService"
   cachedConstructor = null
   newInstanceCallerCache = null
   name = "io.apicurio.registry.client.RegistryService"
   classLoader = {io.quarkus.bootstrap.classloading.QuarkusClassLoader@8592} "QuarkusClassLoader:Quarkus Base Runtime ClassLoader"

Where actual argument then gets this classloader ...

argumentsFromTccl = {java.util.ArrayList@8577}  size = 1
 0 = {io.apicurio.registry.client.CachedRegistryService@8595} ""
  getClass().getClassLoader() = {sun.misc.Launcher$AppClassLoader@6990}

**Comment by geoand** [comment]
I went through this and unfortunately there is no good solution...

Although the code in the custom extension can be hacked to load everything from the TCCL instead of the CL which the methods run, that won't cut it unfortunately because JUnit then validates the returned object against the parameter type (and of course this validation fails because of the different ClassLoaders).

This is somewhat related to #8252.

Perhaps your custom extension could return a proxy (constructed on the regular CL) that then delegates the calls to the real object which is constructed on the TCCL.

**Comment by alesj** [comment]
>> Perhaps your custom extension could return a proxy (constructed on the regular CL) that then delegates the calls to the real object which is constructed on the TCCL.

How would this help?
Wouldn't this still create a mismatch?

**Comment by geoand** [comment]
Ah sorry, yes you are right....

**Comment by alesj** [comment]
I worked around this by introducing Supplier
* https://github.com/Apicurio/apicurio-registry/pull/372

**Comment by geoand** [comment]
You did need to make the changes I suggested in chat, right?

**Comment by alesj** [comment]
Which changes?

I did need to use reflection yes.

See PR.

**Comment by geoand** [comment]
Yeah, that's what I meant. The use of reflection in your extension

**Comment by stuartwdouglas** [comment]
I dont think there is much we can do here without JUnit changes. They are considering adding better support for custom ClassLoader implementations in the next release, so hopefully that happens in a way that is useful to us.

**Comment by geoand** [comment]
This is one that would potentially be fixed by @holly-cummins's WIP
